pylint/extensions/check_raise_docs.py

```diff
@@ -0,0 +1,265 @@
+        \s*     [-=]+   \s*?$          # underline
+        (  .* )                        # section
+    """, re.X | re.S | re.M)
+
+    re_numpy_raise_line = re.compile(r"""
+        \s*  (\w+)                    # type declaration
+    """, re.X)
+
+    def visit_raise(self, node):
+        func_node = node.frame()
+        if not isinstance(func_node, astroid.FunctionDef):
+            return
+
+        excs = _possible_exc_types(node)
+        if not excs:
+            return
+
+        doc = func_node.doc
+        if doc is None:
+            if self.config.accept_no_raise_doc:
+                return
+        elif self.re_sphinx_raise_in_docstring.search(doc) is not None:
+            self._check_sphinx_doc(doc, excs, func_node)
+            return
+        else:
+            match, is_google, re_line = self._establish_section_type(doc)
+            if match is None:
+                # some other documentation style
+                self._add_all_messages(excs, func_node)
+                return
+
+            min_indentation = self._establish_min_section_indentation(
+                match,
+                is_google)
+
+            is_first = True
+            for line in match.group(2).splitlines():
+                if not line.strip():
+                    continue
+                indentation = space_indentation(line)
+                if indentation < min_indentation:
+                    break
+
+                # The first line after the header defines the minimum
+                # indentation.
+                if is_first:
+                    min_indentation = indentation
+                    is_first = False
+
+                if indentation == min_indentation:
+                    # Lines with minimum indentation must contain the beginning
+                    # of a new parameter documentation.
+                    match = re_line.match(line)
+                    if match is None:
+                        break
+
+                    raise_type = match.group(1)
+                    if raise_type in excs:
+                        excs.remove(raise_type)
+                        if not excs:
+                            return
+
+        self._add_all_messages(excs, func_node)
+
+    def _check_sphinx_doc(self, doc, excs, node):
+        """
+        Check a docstring known to be a sphinx docstring for exception types.
+
+        :param doc: The docstring to check.
+        :type doc: str
+
+        :param excs: The exception types to search for.
+        :type excs: list(str)
+
+        :param node: The node to raise warning on.
+        """
+        for match in re.finditer(self.re_sphinx_raise_in_docstring, doc):
+            raise_type = match.group(2)
+            if raise_type in excs:
+                excs.remove(raise_type)
+                if not excs:
+                    return
+
+        self._add_all_messages(excs, node)
+
+    @classmethod
+    def _establish_section_type(cls, doc):
```

Comment:
I put it here because the way it works seemed specific enough to the class to put it in the class rather than module level. I can put it module level if you'd prefer though. I have no major preference either way.